CHANGES.md

```diff
@@ -0,0 +1,94 @@
+# Epochs
+
+Epochs partition the chain of blocks in chunks that can get batched together such that nodes joining the network can rely on repeated checkpoints, rather than downloading and replaying every block.
+
+An epoch starts by increasing the epoch counter by one and resettting the round counter to one. Nodes that are part of the committee which got assigned to this epoch, should propose and vote on blocks. Once an epoch got concluded, the next committee takes over and is responsible for producing and validating blocks.
+
+There are multiple possible ways to initiate the conclusion of an epoch, such as dynamically using system smart contracts or by statically following a hard-coded scheme. To keep things simple, in this implementation, epoch changes happen after a CLI-configurable amount of blocks. Please note that at the moment, there is no consensus on the utilized parameters, which means that all nodes participating in the network need to be started with the same epoch length.
+
+```
+// content of the parameters.json file
+{
+  "consensus": {
+    // ...
+    "epoch_len": 1024
+    // ...
+  }
+}
+```
+
+Note that if the `epoch_len` entry is missing, the nodes do not perform or accept any epoch changes.
+
+```sh
+cargo run --bin node run --parameters=./parameters.json # --keys= ... --store=... --committee=...
+```
+
+The actual epoch change happens by the producer of the block which sets the added property `epoch_concluded` to `true` and thus signals to all other consensus nodes that this particular block has been the last block in this epoch. However, as the implementation does not support dynamically initiating epoch changes, other nodes will refuse to vote for a block at the designated epoch change that does not come with the `epoch_concluded` property set to `true`. Once receiving the proposal by the block proposer, the nodes craft a `Vote` and send the vote the designated proposer of the next block.
```

Comment:
With this solution, a boolean flag indicates whether it's a final block of the epoch or not.

My question is whether this is sufficient information?
Epoch change encompasses two changes:
* epoch number increment
* committee change

I believe you've covered the first point.
But for the 2nd point - how to know which committee has to take over in the next epoch?

Reply:
Regarding the second point: In its current form the implementation expects that if an epoch ends, there will be a committee change as well. However, the committee might consist of the same nodes.

In terms of security, there is indeed a flaw. The passing committee does not specify the upcoming committee. I.e. the nodes could do so by adding some fingerprint of the upcoming committee. Such a fingerprint could look like:
```
committee_fingerprint = hash ( epoch, ordered_publickeys )
```

For the moment, the series of committees is static and it is expected that all nodes have this information available, i.e. as a CLI argument, so the next committee is clear from the context and all honest nodes will refuse blocks from unexpected committees.

Reply:
Or there could be a field in the block/certificate like `next_committee: Option<Committee>` 🙂 
Then you get the hash "for free" as part of the block hash.

If nodes agree on such a block, then we can know for sure that is indeed the next committee.

Reply:
> Or there could be a field in the block/certificate like next_committee: Option<Committee>

Yes, this is what I meant. By doing so, I already gave a way to serialize `Option<Committee>`